dali/python/nvidia/dali/pipeline.py

```diff
@@ -584,13 +659,60 @@ def save_graph_to_dot_file(self, filename, show_tensors = False, show_ids = Fals
             raise RuntimeError("Pipeline must be built first.")
         self._pipe.SaveGraphToDotFile(filename, show_tensors, show_ids, use_colors)
 
+    def set_outputs(self, *outputs):
+        """This function is an alternative way to specify the output data nodes of the pipeline."""
+        self._graph_out = outputs
+
     def define_graph(self):
         """This function is defined by the user to construct the
         graph of operations for their pipeline.
 
         It returns a list of outputs created by calling DALI Operators."""
         raise NotImplementedError
 
+    @staticmethod
+    def _check_data_batch(data, batch_size, layout):
+        if isinstance(data, (list, tuple)):
+            if len(data) != batch_size:
+                raise RuntimeError("The external source callback returned an unexpected batch "
+                "size: {} instead of {}".format(len(data), batch_size))
+            if len(data) > 0:
+                dim = len(data[0].shape)
+                for t in data:
+                    if len(t.shape) != dim:
+                        raise RuntimeErrorError("All tensors in a batch must have the same number of dimensions")
+                if layout != "" and dim != len(layout):
+                    raise RuntimeError("The layout '{}' cannot describe {}-dimensional data".format(layout, dim))
+        else:
+            dim = len(data.shape) - 1
+            if data.shape[0] != batch_size:
+                raise RuntimeError("The external source callback returned an unexpected batch "
+                "size: {} instead of {}".format(data.shape[0], batch_size))
+            if layout != "" and dim != len(layout):
+                raise RuntimeError("The layout '{}' cannot describe {}-dimensional data".format(layout, dim))
+
+    def _run_input_callbacks(self):
+        if self._input_callbacks is None:
+            return
+
+        for group in self._input_callbacks:
+            callback = group[0]._callback
+            callback_out = callback()
+            if group[0]._output_index is not None:
+                for op in group:
+                    data = callback_out[op._output_index]
+                    Pipeline._check_data_batch(data, self._batch_size, op._layout)
+                    self.feed_input(op._name, data, op._layout)
+            else:
+                data = callback_out
+                op = group[0]
+                Pipeline._check_data_batch(data, self._batch_size, op._layout)
+                self.feed_input(op._name, data, op._layout)
```

Comment:
I feel like this is a bit obscure dependency between a wrapper for an Op implementation and the Pipeline implementation. 
Can the group of callbacks be wrapped in some interface? (or at least can we place the comment how does this part behave)

Reply:
The latter - definitely. As for the former - maybe. I'm wondering if and how to move this logic outside Pipeline class.